crates/paymaster-execution/src/tokens/mod.rs

```diff
@@ -0,0 +1,344 @@
+            let unknown_address = Felt::from(0x123u64);
+
+            // When - this will refresh the cache and then look for the unknown token
+            let result = service.normalize_amount(unknown_address, 1000000000000000000).await;
+
+            // Then
+            assert!(result.is_none());
+        }
+
+        #[tokio::test]
+        async fn should_normalize_eth_amount_after_refresh() {
+            // Given
+            let service = TokenService::mainnet();
+            let eth_address = Felt::from_hex_unchecked("0x049d36570d4e46f48e99674bd3fcc84644ddd6b96f7c741b1562b82f9e004dc7");
+            let raw_amount: u128 = 100_000_000_000_000_000; // 0.1 ETH in wei (18 decimals)
+
+            // When - normalize will automatically refresh cache if empty
+            let normalized = service.normalize_amount(eth_address, raw_amount).await;
+            assert!(normalized.is_some(), "ETH token not found in cache");
+            let value = normalized.unwrap();
+            assert!((value - 0.1).abs() < 0.0001, "Expected 0.1, got {}", value);
+
+            // Verify token info
+            let token = service.get(eth_address).await;
+            assert!(token.is_some());
+            let token = token.unwrap();
+            assert_eq!(token.symbol, "ETH");
+            assert_eq!(token.decimals, 18);
+        }
+    }
+
+    mod e2e {
+        use super::*;
+
+        #[tokio::test]
+        async fn should_fetch_tokens_from_mainnet_api() {
+            // Given
+            let service = TokenService::mainnet();
+            let eth_address = Felt::from_hex_unchecked("0x049d36570d4e46f48e99674bd3fcc84644ddd6b96f7c741b1562b82f9e004dc7");
+
+            // When - get will trigger a refresh
+            let result = service.get(eth_address).await;
+
+            // Then
+            assert!(result.is_some(), "Should have fetched ETH token");
+        }
+
+        #[tokio::test]
+        async fn should_fetch_tokens_from_sepolia_api() {
+            // Given
+            let service = TokenService::sepolia();
+            let eth_address = Felt::from_hex_unchecked("0x049d36570d4e46f48e99674bd3fcc84644ddd6b96f7c741b1562b82f9e004dc7");
+
+            // When - get will trigger a refresh
+            let result = service.get(eth_address).await;
+
+            // Then - Sepolia should also have ETH
+            assert!(result.is_some(), "Should have fetched ETH token on Sepolia");
+        }
+
+        #[tokio::test]
+        async fn should_find_common_tokens() {
+            // Given
+            let service = TokenService::mainnet();
+            let eth_address = Felt::from_hex_unchecked("0x049d36570d4e46f48e99674bd3fcc84644ddd6b96f7c741b1562b82f9e004dc7");
+            let strk_address = Felt::from_hex_unchecked("0x04718f5a0fc34cc1af16a1cdee98ffb20c31f5cd61d6ab07201858f4287c938d");
+
+            // When - get() will automatically refresh if cache is empty
+            let eth = service.get(eth_address).await;
```

Comment:
How are we checking that it refreshes if cache is empty ? We are not checking that the cache is empty in the first place